File: backend/src/main/java/com/configuration/account/AccountService.java
package com.configuration.account;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.UUID;

import org.springframework.security.core.context.SecurityContextHolder;

import org.springframework.stereotype.Service;

import com.configuration.account.dto.AccountResponse;
import com.configuration.account.dto.BalanceResponse;
import com.configuration.account.dto.CreateAccountRequest;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AccountService {
    private final AccountRepository accountRepository;

    // ======= POST METHOD ========================================================================
    // ======= Create Account =======
    @Transactional
    public AccountResponse createAccount(UUID userId, CreateAccountRequest request) {
        // String userIdStr =
        // SecurityContextHolder.getContext().getAuthentication().getName();
        // UUID userId = UUID.fromString(userIdStr);

        Account account = Account.builder()
                .userId(userId)
                .accountNumber(generateAccountNumber(request.getAccountType()))
                .accountType(request.getAccountType())
                .accountCategory(request.getAccountType().getCategory())
                .balance(request.getInitialDeposit())
                .status(AccountStatus.ACTIVE)
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        Account saved = accountRepository.save(account);
        return AccountResponse.from(saved);
    }

    // ======= PATCH METHOD ========================================================================
    // ------- Change Balance -------
    @Transactional
    public AccountResponse changeBalance(UUID accountId, BigDecimal amount) {
        // เดี๋ยวถ้ามี token แล้วจะทำการเช็คว่าเป็น admin หรือเป็นเจ้าของบัญชีไหม
        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));

        account.setBalance(account.getBalance().add(amount));
        Account saved = accountRepository.save(account);
        return AccountResponse.from(saved);
    }

    // ------- Delete Account -------
    @Transactional
    public AccountResponse deleteAccount(UUID accountId) {
        // เดี๋ยวถ้ามี token แล้วจะทำการเช็คว่าเป็น admin หรือเป็นเจ้าของบัญชีไหม
        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));
        account.setStatus(AccountStatus.CLOSED);
        Account saved = accountRepository.save(account);
        return AccountResponse.from(saved);
    }

    // ======= GET METHOD ==========================================================================

    public BalanceResponse getAccountBalance(UUID accountId){
        // เดี๋ยวถ้ามี token แล้วจะทำการเช็คว่าเป็น admin หรือเป็นเจ้าของบัญชีไหม
        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));
        return new BalanceResponse(
            account.getAccountNumber(),
            account.getBalance(),
            account.getUpdatedAt()
        );
    }

    public List<AccountResponse> getAllAccounts(){
        // เดี๋ยวระบบ token มาแล้วจะเช็ค user ว่าเป็น admin ไหม
        List<Account> accounts = accountRepository.findAll();
        return accounts.stream()
            .map(AccountResponse::from)
            .toList();
    }

    public AccountResponse getAccountById(UUID accountId){
        // เดี๋ยวถ้ามี token แล้วจะทำการเช็คว่าเป็น admin หรือเป็นเจ้าของบัญชีไหม
        Account account = accountRepository.findById(accountId)
                .orElseThrow(() -> new RuntimeException("Account not found"));
        return AccountResponse.from(account);
    }


    // ======= PRIVATE HELPER ======================================================================

    private String generateAccountNumber(AccountType accountType) {
        // format: BBB-T-XXXXXX
        // BBB = branchNumber (fixed ไปก่อน)
        // T = accountType numberValue
        // XXXXXX = random 6 หลัก

        String branchNumber = "001"; // fixed รอ branch feature
        String typeNumber = accountType.getNumberValue();
        String random = String.format("%06d", (int) (Math.random() * 1000000));

        return branchNumber + typeNumber + random;
    }
}

File: frontend/src/services/accountService.ts
import api from '../lib/axios'
import type { AccountResponse, BalanceResponse, CreateAccountRequest, UserAccountResponse } from '../types/accountType';
import type { PageResponse } from '../types/pageType';


const BASE = "/api/accounts";

export const accountService = {

    // Get /api/accounts
    getAllAccounts: () => api.get<AccountResponse[]>(BASE),

    // Get /api/accounts/admin
    getAccountsPaginated: (page: number, size: number) => 
        api.get<PageResponse<AccountResponse>>(`${BASE}/admin?page=${page}&size=${size}`),

    // Get /api/accounts/{accountId}
    getAccountById: (accountId: string) => api.get<AccountResponse>(`${BASE}/${accountId}`),

    // Get /api/accounts/{accountId}/balance
    getBalanceById: (accountId: string) => api.get<BalanceResponse>(`${BASE}/${accountId}/balance`),

    // Get /api/accounts/user
    getUserAccount: (page: number = 0, size: number = 10) => 
        api.get<PageResponse<UserAccountResponse>>(`${BASE}/user?page=${page}&size=${size}`),

    // Post /api/accounts
    createAccount: (data: CreateAccountRequest) => api.post<AccountResponse>(BASE, data),

    // Patch /api/accounts/{accountId}/balance?amount={amount}
    addBalanceToAccount: (accountId: string, amount: number) => api.patch<AccountResponse>(`${BASE}/${accountId}/balance?amount=${amount}`),
    
    // Patch /api/accounts/{accountId}/status
    changeAccountStatus: (accountId: string, status: string) => api.patch<AccountResponse>(`${BASE}/${accountId}/status`, {"status": status}),
    
    // Patch /api/accounts/{accountId}/type
    changeAccountType: (accountId: string, accountType: string) => api.patch<AccountResponse>(`${BASE}/${accountId}/type`, {"accountType": accountType}),

    // Delete /api/accounts/{accountId}
    deleteAccount: (accountId: string) => api.delete<AccountResponse>(`${BASE}/${accountId}`),

}